File: ros2_ws/src/localization/localization/dead_reckoning.py
import rclpy
from rclpy.node import Node
from cascade_msgs.msg import SensorReading
from geometry_msgs.msg import PoseStamped
from geometry_msgs.msg import Quaternion
from message_filters import ApproximateTimeSynchronizer, Subscriber
import tf2_ros
from tf2_ros import TransformListener, Buffer
import tf2_geometry_msgs
import geometry_msgs.msg
import time
import math

class DeadReckoningNode(Node):
    def __init__(self):
        super().__init__ ("dead_reckoning_node")
        queue_size=20
        acceptable_delay=0.01 #this is how many seconds of difference we allow between the 2 subscriptions before theyre considered not matching
        tss = ApproximateTimeSynchronizer(
            [Subscriber(self, SensorReading, "/PID/yaw/actual"),
                Subscriber(self, SensorReading, "/PID/pitch/actual"),
                Subscriber(self, SensorReading, "/PID/roll/actual"),
                Subscriber(self, SensorReading, "/PID/surge/actual"),
                Subscriber(self, SensorReading, "/PID/sway/actual"),
                Subscriber(self, SensorReading, "/PID/heave/actual"),
            ],
            queue_size,
            acceptable_delay)
        tss.registerCallback(self.synced_callback)
        self.publisher_ = self.create_publisher(PoseStamped, '/pose', 10)
        self.current_x=0.0
        self.current_y=0.0
        self.current_z=0.0
        self.last_time = time.perf_counter()
        self.tf_buffer = Buffer()
        self.tf_listener = TransformListener(self.tf_buffer, self)

    def euler_to_quaternion(self, roll, pitch, yaw):
        """Convert Euler angles to quaternion."""
        qx = math.sin(roll / 2) * math.cos(pitch / 2) * math.cos(yaw / 2) - math.cos(roll / 2) * math.sin(pitch / 2) * math.sin(yaw / 2)
        qy = math.cos(roll / 2) * math.sin(pitch / 2) * math.cos(yaw / 2) + math.sin(roll / 2) * math.cos(pitch / 2) * math.sin(yaw / 2)
        qz = math.cos(roll / 2) * math.cos(pitch / 2) * math.sin(yaw / 2) - math.sin(roll / 2) * math.sin(pitch / 2) * math.cos(yaw / 2)
        qw = math.cos(roll / 2) * math.cos(pitch / 2) * math.cos(yaw / 2) + math.sin(roll / 2) * math.sin(pitch / 2) * math.sin(yaw / 2)
 
        return Quaternion(x=qx, y=qy, z=qz, w=qw)

    def synced_callback(self, yaw_msg, pitch_msg, roll_msg, surge_msg, sway_msg, heave_msg):
        pose_msg=PoseStamped()

        pose_msg.header.stamp=self.get_clock().now().to_msg() 
        delta_t = time.perf_counter() - self.last_time; 
        self.last_time=time.perf_counter();

        quaternion = self.euler_to_quaternion(math.radians(roll_msg.data), math.radians(pitch_msg.data), math.radians(yaw_msg.data))

        # Create a TransformStamped message for the body-to-world transformation
        transform = geometry_msgs.msg.TransformStamped()
        transform.header.stamp = pose_msg.header.stamp
        transform.header.frame_id = 'world'
        transform.child_frame_id = 'base_link'  # Your body frame

        transform.transform.rotation = quaternion

        # Optionally, set the translation part if you need to track position
        transform.transform.translation.x = self.current_x
        transform.transform.translation.y = self.current_y
        transform.transform.translation.z = self.current_z
        
        br = tf2_ros.TransformBroadcaster(self)
        br.sendTransform(transform)
        
        vel_body = geometry_msgs.msg.Vector3()
        vel_body.x = surge_msg.data
        vel_body.y = sway_msg.data
        vel_body.z = heave_msg.data

        # Convert Vector3 to a stamped vector so that tf2 can handle it
        vel_body_stamped = geometry_msgs.msg.Vector3Stamped()
        vel_body_stamped.vector = vel_body
        vel_body_stamped.header.frame_id = 'base_link'
        vel_body_stamped.header.stamp = pose_msg.header.stamp

        # Transform velocity into the world frame
        try:
            # Transform the velocity to the world frame
            vel_world = self.tf_buffer.transform(vel_body_stamped, 'world', timeout=rclpy.duration.Duration(seconds=0.05))
            vel_world = vel_world_stamped.vector
            self.current_x += vel_world.x * delta_t
            self.current_y += vel_world.y * delta_t
            self.current_z += vel_world.z * delta_t
            self.get_logger().info(f'Transformed velocity: {transformed_vel.vector}')
        except Exception as ex:
            self.get_logger().warn(f'Failed to transform velocity: {str(ex)}')

        pose_msg.pose.position.x=self.current_x
        pose_msg.pose.position.y=self.current_y
        pose_msg.pose.position.z=self.current_z
        pose_msg.pose.orientation=quaternion

        self.publisher_.publish(pose_msg)

def main(args=None):
    rclpy.init(args=args)
    node = DeadReckoningNode()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: ros2_ws/src/navigation/src/dead_reckoning.cpp
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/pose_stamped.hpp"
#include "geometry_msgs/msg/quaternion.hpp"
#include "cascade_msgs/msg/sensor_reading.hpp"
#include "sensor_msgs/msg/imu.hpp"
#include <tf2/LinearMath/Quaternion.h>
#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>
#include <chrono>

typedef unsigned int uint32;
using std::placeholders::_1;

class DeadReckoningNode : public rclcpp::Node{
    public:
        double x,y,z,roll=0,pitch=0,yaw=0;
        bool gotFirstTime=false;
        std::chrono::time_point<std::chrono::high_resolution_clock> last_time;

	    DeadReckoningNode() : Node("dead_reckoning_node") {
            rclcpp::QoS qos_profile(10);  // Create QoS profile with history depth 10
            qos_profile.reliability(rclcpp::ReliabilityPolicy::BestEffort);

		    gyro_subscription = this->create_subscription<sensor_msgs::msg::Imu>("/camera/camera/gyro/sample", qos_profile, std::bind(&DeadReckoningNode::imu_callback, this, _1));

		    roll_publisher = this->create_publisher<cascade_msgs::msg::SensorReading>("PID/roll/actual", 10);
		    pitch_publisher = this->create_publisher<cascade_msgs::msg::SensorReading>("PID/pitch/actual", 10);
	    	yaw_publisher = this->create_publisher<cascade_msgs::msg::SensorReading>("PID/yaw/actual", 10);
	    	pose_publisher = this->create_publisher<geometry_msgs::msg::PoseStamped>("/pose", 10);
	    }

	private:
	void imu_callback(const sensor_msgs::msg::Imu &msg) {
        if(!gotFirstTime){
            last_time = std::chrono::high_resolution_clock::now();
            gotFirstTime=true;
            return;
        }
		std::chrono::duration<double> duration = std::chrono::high_resolution_clock::now() - last_time;
        double dt = duration.count();
        last_time = std::chrono::high_resolution_clock::now();
		
		pitch  += msg.angular_velocity.x * dt;
		yaw -= msg.angular_velocity.y * dt;
        roll   -= msg.angular_velocity.z * dt;
        //yaw and roll are negative to convert from D455 coordinate plane to standard

		auto roll_msg  = cascade_msgs::msg::SensorReading();
		auto pitch_msg = cascade_msgs::msg::SensorReading();
		auto yaw_msg   = cascade_msgs::msg::SensorReading();
		auto pose_msg   = geometry_msgs::msg::PoseStamped();

        roll_msg.data=roll*180.0/3.141592653589793238463;
        yaw_msg.data=yaw*180.0/3.141592653589793238463;
        pitch_msg.data=pitch*180.0/3.141592653589793238463;

        tf2::Quaternion q;
        q.setRPY(roll, pitch, yaw);

        pose_msg.pose.orientation=tf2::toMsg(q);

        roll_msg.header.stamp=msg.header.stamp;
        pitch_msg.header.stamp=msg.header.stamp;
        yaw_msg.header.stamp=msg.header.stamp;
        pose_msg.header.stamp=msg.header.stamp;
        pose_msg.header.frame_id="map";

		pitch_publisher -> publish(pitch_msg);
		roll_publisher  -> publish(roll_msg);
		yaw_publisher   -> publish(yaw_msg);
        pose_publisher  -> publish(pose_msg);
	}
	rclcpp::Subscription<sensor_msgs::msg::Imu>::SharedPtr gyro_subscription;
	rclcpp::Publisher<cascade_msgs::msg::SensorReading>::SharedPtr roll_publisher;
	rclcpp::Publisher<cascade_msgs::msg::SensorReading>::SharedPtr pitch_publisher;
	rclcpp::Publisher<cascade_msgs::msg::SensorReading>::SharedPtr yaw_publisher;
	rclcpp::Publisher<geometry_msgs::msg::PoseStamped>::SharedPtr pose_publisher;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<DeadReckoningNode>());
  rclcpp::shutdown();


  return 0;
}
